test/test_player.py

```diff
@@ -63,6 +89,380 @@ def test_backslash_in_arg(self):
+        """
+
+        while True:
+            if terminator in self.buf:
+                line, self.buf = self.buf.split(terminator, 1)
+                line += terminator
+                return line
+            self.sock.settimeout(1)
+            data = self.sock.recv(bufsize)
+            if data:
+                self.buf += data
+            else:
+                line = self.buf
+                self.buf = b''
+                return line
+
+
+def start_beets(*args):
+    import beets.ui
+    beets.ui.main(list(args))
+
+
+def implements(commands, expectedFailure=False):  # noqa: N803
+    def _test(self):
+        with self.run_bpd() as client:
+            response = client.send_command('commands')
+        implemented = response.data['command']
+        self.assertEqual(commands.intersection(implemented), commands)
+    return unittest.expectedFailure(_test) if expectedFailure else _test
+
+
+class BPDTest(unittest.TestCase, TestHelper):
+    def setUp(self):
+        self.setup_beets(disk=True)
+        self.load_plugins('bpd')
+        self.item1 = self.add_item(title='Track One Title',
+                                   album='Album Title', artist='Artist Name',
+                                   track=1)
+        self.item2 = self.add_item(title='Track Two Title',
+                                   album='Album Title', artist='Artist Name',
+                                   track=2)
+        self.lib.add_album([self.item1, self.item2])
+
+    def tearDown(self):
+        self.teardown_beets()
+        self.unload_plugins()
+
+    @contextmanager
+    def run_bpd(self, host='localhost', port=9876, password=None,
+                           do_hello=True):
+        """ Runs BPD in another process, configured with the same library
+        database as we created in the setUp method. Exposes a client that is
+        connected to the server, and kills the server at the end.
+        """
+        # Create a config file:
+        config = {
+                'pluginpath': [py3_path(self.temp_dir)],
+                'plugins': 'bpd',
+                'bpd': {'host': host, 'port': port},
+        }
+        if password:
+            config['bpd']['password'] = password
+        config_file = tempfile.NamedTemporaryFile(
+                mode='wb', dir=py3_path(self.temp_dir), suffix='.yaml',
+                delete=False)
+        config_file.write(
+                yaml.dump(config, Dumper=confit.Dumper, encoding='utf-8'))
+        config_file.close()
+
+        # Fork and launch BPD in the new process:
+        args = (
+            '--library', self.config['library'].as_filename(),
+            '--directory', py3_path(self.libdir),
+            '--config', py3_path(config_file.name),
+            'bpd'
+        )
+        server = mp.Process(target=start_beets, args=args)
+        server.start()
+
+        # Wait until the socket is bound:
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```

Comment:
I think this socket creation has to go *inside* the loop. At least on my system (macOS 10.14.4), if the connection fails *once* when the server is not ready, then subsequent attempts to call `self.sock.connect_ex` return error 22 (invalid argument).

Reply:
Ah, ok. It makes more sense that it works that way, thanks.